_posts/2017-09-04-Secure-your-architecture-part1.md

```diff
@@ -0,0 +1,427 @@
+        return new JwtTokenStore(accessTokenConverter());
+    }
+
+    @Bean
+    @Primary
+    public DefaultTokenServices tokenServices() {
+        DefaultTokenServices defaultTokenServices = new DefaultTokenServices();
+        defaultTokenServices.setTokenStore(tokenStore());
+        return defaultTokenServices;
+  }
+}
+
+{% endhighlight %}
+
+### Common mistakes with keys
+So the UAA signs the token with a private key and verifies it with a public key. 
+The problem that occurs is that every microservice would need the public key to connect with the UAA server for verification on every request.
+As you can see, we don't want every microservice to be dependant of the UAA server to be available at startup, in testing or CI. 
+The solution to this is to disable exposure of your microservices to the outer network and handle only incoming traffic via the reverse proxy/load balancer (Zuul).
+Zuul will verify the token as a trustworthy client of the UAA server and will propagate the token to the downstreamed services.
+The microservice will handle the token as trustworthy since the token is verified.
+
+
+### Securing API endpoints
+At last we're going to secure our resources.
+Spring Security gives us a variety of tools to secure at class/method level.
+Let me point you out the most used one.  <br />
+To enable method security add `@EnableGlobalMethodSecurity(prePostEnabled = true)` to your configuration. 
+
+#### Authority 
+For the authorization, Spring Security provides us authorities, extracted from the JWT, the authorities are placed inside a principal.
+This principal will be used throughout the existing security context of your application. 
+Now these authorities can be used for authorizing methods. 
+It provides us method based access control where you have the ability to use SPEL expressions as an authorization mechanism.
+There is a lot of options to use for method security but for now we will see the most common ones.
+
+For the full list: [Spring Expression Language](https://docs.spring.io/spring-security/site/docs/3.0.x/reference/el-access.html)
+
+##### @PreAuthorize
+Most commonly used, PreAuthorize will decide whether a method can actually be invoked or not.
+>When a user logs in and you want the user to only access his detail information, you can use the PreAuthorize annotation to accept USER roles with a specific expression to see if it is the same user calling the endpoint.
+>If you like the ADMIN role to access everyone's details, you can just add the hasRole expression.
+
+ {% highlight java %}
+@PreAuthorize("(authentication.principal.uuid == #uuid.toString()) or hasRole('ADMIN')")
+User findByUuid(@Param("uuid") UUID uuid);
+ {% endhighlight %}
+
+##### @PostAuthorize
+Less commonly, you may wish to perform access-control check after the method has been invoked.
+The returnObject is the returned value of that method.
+>A USER role can only update his own details and not from everyone else, but an ADMIN role can.
```

Comment:
Wat gebeurt er als ik een ander user heb proberen te saven? Wordt deze dan toch opgeslagen of wordt de wijziging teruggedraaid? Misschien is een ander voorbeeld beter: kijken of de opgehaalde user inderdaad opgehaald mag worden door de user die het probeert te doen (of iets dergelijks)